\documentclass[a4paper, 11pt]{article}

%% Language and font encodings
\usepackage[english]{babel}
\usepackage[utf8x]{inputenc}
\usepackage[T1]{fontenc}

%% Sets page size and margins
\usepackage[a4paper,top=3cm,bottom=2cm,left=3cm,right=3cm,marginparwidth=1.75cm]{geometry}

%% Useful packages
\usepackage{amsfonts,amssymb,amsmath,amscd,amsthm,latexsym}
\usepackage{tikz,graphicx,enumerate}
\usepackage{natbib}
\usepackage[colorinlistoftodos]{todonotes}
\usepackage[colorlinks=true, allcolors=blue]{hyperref}

\newcommand{\cst}{\text{cst}}
\newcommand{\Esp}{\mathbb{E}}
\newcommand{\Ibb}{\mathbb{I}}
\newcommand{\Hcal}{\mathcal{H}}
\newcommand{\Ncal}{\mathcal{N}}
\newcommand{\Pcal}{\mathcal{P}}
\newcommand{\Tcal}{\mathcal{T}}
\newcommand{\Ucal}{\mathcal{U}}
\newcommand{\Zcal}{\mathcal{Z}}
\newcommand{\Dx}{\Delta x}
\newcommand{\Tt}{\widetilde{T}}
\newcommand{\xb}{\overline{x}}
\newcommand{\Dxb}{\Delta \overline{x}}
\newcommand{\DY}{\Delta Y}

\title{Combining tests for data integration}
\author{}

\begin{document}
\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Problem and notations}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\paragraph{Problem.} Detect pairs of gene / CpG probes being both associated with the status, that is where both a differential expression and a differential methylation are observed.

\paragraph{Notations.}
\begin{itemize}
 \item $i = 1 ... n$ patients
 \item $j = 0, 1$ sub-sample status (0 = control, 1 = tumor) in a patient
 \item $k = 1 ... p$ genes
 \item $\ell = 1 ... L_k$ locus of methylation sites (for gene $k$)
 \item $Y_{ijk} =$ (log-)expression of gene $k$ in sub-sample $j$ from patient $i$
 \item $x_{ijk\ell} =$ methylation at locus $\ell$ for gene $k$ in sub-sample $j$ from patient $i$
\end{itemize}

\paragraph{Tests.} For each gene $k$, we may define a test statistic $T^E_k$ (function of the $(Y_{ijk})$) to test 
$$
H_{0k}^E = \left\{\text{gene $k$ has the same expression level in both status}\right\}.
$$
Knowing the null (i.e. under $H^E_0$) distribution $F^E_{0k}$ of $T^E_k$ we can get the p-value $P^E_k$. \\
Similarly, we can define $T^M_{k\ell}$, $H_{0k\ell}^E$, $F^M_{0k\ell}$ and $P^M_{k\ell}$ for the $\ell$th methylation site of gene $k$.

\paragraph{Aims.} We are actually interested in couple $(k, \ell)$ such that both 
$H_{0k}^E$ and $H_{0k\ell}^E$ are rejected, meaning that we are interested to test
$$
H_{0k\ell} = H_{0k}^E \cup H_{0k\ell}^M
\qquad \text{versus} \qquad 
H_{1k\ell} = H_{1k}^E \cap H_{1k\ell}^M.
$$

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Resampling strategy}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\paragraph{Problem.} Defining the distribution under the {\sl union} of null hypotheses is not standard. Even resampling is not standard as we need to resample only from this union of nulls.

The proposed strategy consists in
\begin{enumerate}[($a$)]
 \item For each hypothesis, compute the conditional probability of $H_0$ and $H_1$ given the observed statistic
 \item Sample among all the pairs $(k, \ell)$, using the conditional probabilities as weights (to ensure to sample from the union of null hypotheses) to get an estimate of the distribution under the $H_0$.
 \item Compute the p-value according to this estimated null distribution.
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Proposed procedure}

\paragraph{Step $(a)$: Semi-parametric mixture.} We use the semi-parametric approach introduced by \cite{RBD07}. More specifically, we fit the following mixture model to the probit-transform expression p-values $\Tt^E_k$:
$$
\Tt^E_k := -\Phi^{-1}(T^E_k) \sim \pi^E_0 \Ncal(0, 1) + (1 - \pi^E_0) F^E_1
$$
where $\Phi$ stands for the cdf of the standard normal distribution. This semi-parametric approach provides estimates of both the proportion $\pi^E_0$ and the alternatif distribution $F^E_1$. A by product of this approach is the conditionnal probability
$$
\tau^E_k = \Pr\left\{H^E_{1k} \mid \Tt^E_k\right\} = 1 - \Pr\left\{H^E_{0k} \mid \Tt^E_k\right\}.
$$
Test statistics $\Tt^M_{k\ell}$, estimates of $\pi^M_0$ and $F^M_1$ and conditionnal probabilities $\tau^M_{k\ell}$ can be derived for the methylation tests as well.

\paragraph{Step $(b)$: Resampling.} With each pair $(k, \ell)$, we associate the probability
$$
\tau_{k\ell} = \tau^E_k \tau^M_{k\ell} \simeq \Pr\left\{H_{1k\ell} \mid \Tt^E_k,  \Tt^M_{k\ell}\right\}
$$
and sample $B$ pairs $(k_b, \ell_b)$ among all pairs $(k, \ell)$ with replacement with probabilities $1 - \tau_{k_b\ell_b}$, as we want to sample under $H_{0k\ell} = H^E_{0k} \cup H^M_{0k\ell}$. 

\paragraph{Step $(c)$: p-value.} For a given joint statistic 
$$
T_{k\ell} = t(\Tt^E_k, \Tt^M_{k\ell})
$$
e.g.
$$
t(\Tt^E_k, \Tt^M_{k\ell}) = | \Tt^E_k \Tt^M_{k\ell} |
\qquad \text{or} \qquad
t(\Tt^E_k, \Tt^M_{k\ell}) = \min(\Tt^E_k, \Tt^M_{k\ell}).
$$
we define the p-value
$$
P_{k\ell} = B^{-1} \sum_b \Ibb\{t(\Tt^E_{k_b}, \Tt^M_{k_b\ell_b}) \geq T_{k\ell}\}.
$$

\paragraph{Improving the sampling.} Steps $(b)$ and $(c)$ are somewhat inefficient to estimate  the tail probability as they mostly sample in the center of the null distribution. To improve the efficiency, we may define a weight $\omega_{k\ell}$, being an increasing function of $t(\Tt^E_k, \Tt^M_{k\ell})$, such as
$$
\omega_{k\ell} = \text{rank}(t(\Tt^E_k, \Tt^M_{k\ell}))
$$
and revise the steps as
\begin{description}
 \item[Step $(b')$:] Sample $B$ pairs $(k_b, \ell_b)$ with respective probabilities $\omega_{k\ell} \tau_{k\ell}$.
 \item[Step $(c')$:] Estimate the p-value as
 $$
 P_{k\ell} = \sum_b \omega_{k\ell}^{-1} \Ibb\{t(\Tt^E_{k_b}, \Tt^M_{k_b\ell_b}) \geq T_{k\ell}\} \left/ \sum_b \omega_{k\ell}^{-1} \right.
 $$
\end{description}

\paragraph{Tristan's remark.} If all pairs $(k, \ell)$ were resampled, the procedure described in step ($a$), ($b$) and ($c$) would simply yield
$$
P_{k\ell} 
= \sum_{(k', \ell')} (1-\tau_{k'\ell'}) \Ibb\{t(\Tt^E_{k'}, \Tt^M_{k'\ell'}) \geq T_{k\ell}\}
\left/ \sum_{(k', \ell')} (1-\tau_{k'\ell'}) \right.
$$
which can be computed straightforwardly from the observed data by simply sorting the observed $T_{k\ell}$.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Remarks}

\paragraph{Pros.} 
\begin{enumerate}
 \item The procedure is agnostic with respect to the original test statistics $T^E$ and $T^M$, which can therefore be different, provided their respective null distributions are known. Their transformation into $\Tt^E$ and $\Tt^M$ makes comparable by essence.
 \item The procedure holds for any joint test statistic $T$, not only for the product or the sum, as proposed here.
 \item The procedure easily generalizes to more that two joint tests, that is to more that two genomic signals.
\end{enumerate}

\paragraph{Cons.} 
\begin{enumerate}
 \item It is not clear if the procedure is a genuine test procedure, because of the non-standard for of '$H_0$'.
 \item The procedure strongly relies on the estimates of $F^E_1$ and $F^M_1$.
 \item The definition of the probability $\tau_{k\ell}$ relies on an implicit independence assumption, which does not necessarily make sense for out purpose.
\end{enumerate}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage
\section{Alternative formulation of the problem}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Definitions.}
\begin{itemize}
 \item Let us consider $Q$ different types of measurements collected on a series of individuals ($i = 1 \dots n$) and a series of loci ($k = 1 \dots m$)
 \item Let $Y^q_{ik}$ denote the response of type $q$ measured on individual $i$ at loci $k$.
 \item Let us define the null hypothesis 
 $$
 H^q_{0k} = \{Y^q_{ik} \text{ has the same distribution in two subgroups of individuals}\}.
 $$
 \item Let $P^q_k$ denote the $p$-value associated with $H^q_{0k}$ and 
 $$
 \Tcal_k = (T^q_k)_{1 \leq q \leq Q}
 $$
 \item Let us define the union null hypothesis
 $$
 \Hcal_{0k} = \bigcup^Q_{q=1} H^q_{0k}.
 $$
\end{itemize}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Union-intersection test.} According to Lehmann, define 
$$
\Pcal_k = \max_q P^q_k.
$$
Enables to control FWER over $k$ via Bonferroni, but not FDR control as the distribution of $\Pcal_0$ under $\Hcal_0$ is unknown (because $\Hcal_0$ is composite).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Classification approach.} 
Our aim is to estimate the union false discovery probability
$$
\tau_k = 1 - \Pr\{\Hcal_{0k} \mid \Tcal_k\}.
$$
To this aim, we assume that each $H^q_{0k}$ is simple so that the marginal distribution of $P^q_k$ under $H^q_{0k}$ is $\Ucal_{[0, 1]}$.
% and one may evaluate the conditional probability
% $$
% \tau^q_k = 1 - \Pr\{H^q_{0k} \mid T^q_k\}.
% $$

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Classification model}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Following a local-FDR ($\ell FDR$) approach, we assume that, for each $q$, the $p$-values $\{P^q_k\}_{1 \leq k \leq m}$ are iid, each arising from the mixture model:
\begin{equation} \label{eq:MixturePk}
 P^q_k \sim \pi_0^q F^q_0 + (1-\pi_0^q) F^q_1.
\end{equation}
We further denote $f^q_0$ (resp. $f^q_1$) the density of $F^q_0$ (resp. $F^q_1$). Obviously $F^q_0 = \Ucal_{[0, 1]}$, so $f^q_0 \equiv f_0 \equiv 1$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Union null hypothesis.} 
$\Hcal_{0k}$ includes $2^Q -1$ configurations. Let us define the binary variable
$Z^q_k = 0$ if $H^q_{0k}$ holds and 1 otherwise. Let $\Zcal_k$ denote the configuration of loci $k$: $\Zcal_k = (Z^q_k)_{1 \leq q \leq Q}$. $\Hcal_{0k}$ includes all configurations except $\Zcal_k = (1, \dots 1)$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Joint conditional distribution of the $p$-values.} We assume that the marginal $p$-values arising from a null distribution are independent. So
$$
f_{\Zcal_k}(\Pcal_k) = \prod_{k: Z^q_k=0} f_0(P^q_k) \times f_{1;\Zcal_k}\left((P^q_k)_{q: Z^q_k=1}\right).
$$
In the special case where $Q=2$, we have $\Pcal_k = (P^1_k, P^2_k)$ and
\begin{align*}
  f_{(0,0)}(\Pcal_k) & = f_0(P^1_k) f_0(P^2_k) \\
  f_{(0,1)}(\Pcal_k) & = f_0(P^1_k) f_1(P^2_k) \\
  f_{(1,0)}(\Pcal_k) & = f_1(P^1_k) f_0(P^2_k) \\
  \text{but} \qquad f_{(1,1)}(\Pcal_k) & = f_{11}(P^1_k, P^2_k) & & \text{(no restriction).}
\end{align*} 
so the marginal distributions $f^1_1$ and $f^2_1$ are sufficient to fully define the distribution of $\Pcal_k$ under $\Hcal_{0k}$. \\
In the case where $Q=3$, we have
\begin{align*}
  f_{(0,0,0)}(\Pcal_k) & = \\
  f_{(0,0,1)}(\Pcal_k) & = \\
  \dots 
\end{align*}
and the marginal distributions $f^q_1$ are not sufficient to define $f_{\Zcal_k}(\Pcal_k)$


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Marginal probability of a configuration.} We denote
$$
\rho(z) = \Pr\{\Zcal_k = z\}.
$$
A possible option is
$$
\rho(z) = \prod_q (\pi^q_0)^{1-z^q} (\pi^q_1)^{z^q}.
$$

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Inference}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Marginal mixture distributions.} For each $q$, we can fit the mixture model \eqref{eq:MixturePk} according to \cite{RBD07} to get estimates of $\pi^q_0$ and $f^q_1$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Sampling from the joint non-null distribution.} 


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{Last remarks.} 
\begin{itemize}
 \item Plus de calcul de $p$-value car classif
 \item Pas d'autre idee que le produit des $\tau^q$ pour le poids dans l'estimation de $f_{1, \Zcal}$ (probl√®me seulement pour plus de 2 tests)
 \item Espoir de calcul efficace en se reposant sur le tri des observations $k$ en fonction de la stat $\Pcal_k = \max_q P^q_k$.
\end{itemize}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\paragraph{27 / 11 / 2019 : alternative union $p$-value.} 
\begin{align*}
	P(P_{\max} < p \mid \Hcal_0) 
	& = \left(\pi_{00} P(P_{\max} < p \mid H_{00}) + \pi_{01} P(P_{\max} < p \mid H_{01}) + \pi_{10} P(P_{\max} < p \mid H_{10}) \right) / (1- \pi_{11}) \\
	& = \left(
	\pi_{00} P(P^1 < p \mid H^1_0) P(P^2 < p \mid H^2_0) 
	+ \pi_{01} P(P^1 < p \mid H^1_0) P(P^2 < p \mid H^2_1) 
	+ \pi_{10} P(P^1 < p \mid H^1_1) P(P^2 < p \mid H^2_0) 
	\right) / (1- \pi_{11}) \\
	& = \left(
	\pi_{00} p^2 
	+ \pi_{01} p F_1^2(p)
	+ \pi_{10} p F_1^1(p)
	\right) / (1- \pi_{11}) \\
\end{align*}
where $F_1^q$ can be estimated as
$$
\widehat{F}^q_1(p) = \frac{\sum_{i: P^q_i < p} \tau_i^q}{\sum_{i} \tau_i^q}
$$


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliographystyle{plain}
\bibliography{Reference}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\end{document}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
